includes/class-wc-admin-reports-segmenting.php

```diff
@@ -342,17 +342,31 @@ protected function set_all_segments() {
 			$segment_objects = wc_get_products( $args );
 
 			foreach ( $segment_objects as $segment ) {
-				$id = $segment->get_id();
-				$segments[] = $id;
+				$id                    = $segment->get_id();
+				$segments[]            = $id;
 				$segment_labels[ $id ] = $segment->get_name();
 			}
 		} elseif ( 'category' === $this->query_args['segmentby'] ) {
-			$categories = get_categories(
-				array(
-					'taxonomy' => 'product_cat',
-				)
-			);
-			$segments   = wp_list_pluck( $categories, 'cat_ID' );
+			$where_clause = "WHERE taxonomy = 'product_cat'";
+			if ( 0 < count( $this->query_args['categories'] ) ) {
+				$category_ids  = implode( ',', $this->query_args['categories'] );
+				$where_clause .= " AND {$wpdb->prefix}terms.term_id IN ({$category_ids})";
+			}
+			$categories = $wpdb->get_results(
+				"
+				SELECT {$wpdb->prefix}terms.term_id as catogory_id, name 
+				FROM {$wpdb->prefix}term_taxonomy 
+					LEFT JOIN {$wpdb->prefix}terms on {$wpdb->prefix}term_taxonomy.term_id = {$wpdb->prefix}terms.term_id
+				{$where_clause}
+				",
+				ARRAY_A
+			); // WPCS: cache ok, DB call ok, unprepared SQL ok.
+
+			foreach ( $categories as $category ) {
+				$id                    = $category['catogory_id'];
+				$segments[]            = $id;
+				$segment_labels[ $id ] = $category['name'];
+			}
```

Comment:
`segment_label` was always `NULL` here, preventing the chart from functioning. Also, categories were not being filtered. I could not get `get_categories` to work, so I ended up writing the query.

Is there a better way?

Reply:
Try `wp_list_pluck( $categories, 'term_id' );`.